### ISSUE TITLE
Debounce

### ISSUE THREAD

**Issue by vsuharnikov** [albertlauncher/python#150]
# Issue

Now each entered character triggers `Plugin.handleQuery`, even a user doesn't want to get an intermediate result (see a scenario below).
Also, a plugin does a lot of unnecessary work during the input, especially if a plugin runs a complex task.

A scenario:
1. A user wants to get a translation of a long word
2. `Plugin.handleQuery` is triggered on each character
3. So the plugin makes HTTP requests to google translate on each update, for example: p, pn, pne, ..., pneumonoultramicroscopicsilicovolcanoconiosis

# Solution 1 - Simple, less responsive

The first approach is to call `time.sleep` (see [googletrans](https://github.com/albertlauncher/python/blob/master/googletrans/__init__.py#L48)).
`time.sleep` blocks the thread, so the input becomes less responsive and you have to wait a delay before the entered symbol shows.

# Solution 2 - Complex

The another approach is to use `asyncio` and `threading`:
```python
import asyncio
import threading

class Plugin(QueryHandler):
    is_working = True  # For a background task, that makes the EventLoop alive
    query = None  # The current query
    thead = None  # We need this to move work to a background
    loop = None # To submit tasks and cancel 
    curr_task = None  # To cancel when a new symbol is entered and we haven't yet made the job

    def initialize(self):
        self.loop = asyncio.get_event_loop()
        self.thread = threading.Thread(target=self.runLoop)
        self.thread.start()

    def finalize(self):
        self.is_working = False
        if self.curr_task:
            self.curr_task.cancel()
        self.thread.join()

    def handleQuery(self, query):
        if self.curr_task:
            self.curr_task.cancel()
        self.query = query
        self.curr_task = self.loop.create_task(self.debounced())

    # Internal

    async def debounced(self):
        await asyncio.sleep(0.5)
        if not self.query.isValid:
            return

        # <-------------------------- Your code

        if self.query.isValid:
            self.query.add()

    def runLoop(self):
        self.loop.run_until_complete(self.watchLoop())

    async def watchLoop(self):
        while self.is_working:
            await asyncio.sleep(1)
```

It works good and could be improved further (close the thread on idle?). But it requires a ton of boilerplate as you see. 
Also, it doesn't sound good, if each plugin has own thread and even loop.

# Solution 3 - Native debounce

To support this functionality in the core. It would be great if a developer could choose a suitable value, because 
some plugins do a job locally and fast (e.g. a hash calculation). So they doesn't need this functionality.
```python
class Plugin(QueryHandler):
  def debounce(self):
    return 0.2 # no debounce (0) by default 
```

# Solution 4 - ?

Probably, there is a simpler solution. Any ideas?

**Comment by ManuelSchneid3r** [comment]
Where necessary solution 1 is applied. If there's stuff to do I want albert to do it instantly. requests are a problem there because they are often rate limited. otherwise otherwise fire and discard would be fine. The app cant 🔮 if your query is finished yet. to avoid  unnecesary computation we have isValid(). time is an issue that's why there are triggered queryhandlers. global queries can't wait for network responses.

Im not exactly sure what you expect, because I don't see the problem. can you elaborate on this?

**Comment by ManuelSchneid3r** [comment]
e.g. 

https://github.com/albertlauncher/python/blob/master/arch_wiki/__init__.py#L42-L46

https://github.com/albertlauncher/python/blob/master/aur/__init__.py#L59

https://github.com/albertlauncher/python/blob/master/googletrans/__init__.py#L47

https://github.com/albertlauncher/python/blob/master/wikipedia/__init__.py#L73

**Comment by ManuelSchneid3r** [comment]
Asyncio has a point when trying to workaround the global interpreter lock. but here i dont see in which aspect the asyncio solution should be better.

**Comment by ManuelSchneid3r** [comment]
> blocks the thread

btw queries are multithreaded

**Comment by vsuharnikov** [comment]
> Im not exactly sure what you expect, because I don't see the problem. can you elaborate on this?

The first solution works, but I see UI freezes when typing in the textbox. 

> Asyncio has a point when trying to workaround the global interpreter lock.

Exactly! :)

> but here i dont see in which aspect the asyncio solution should be better.

So, here are demos:

# time.sleep demo

```python
# ~/.local/share/albert/python/plugins/sleep-demo/__init__.py
from albert import *
import time

md_iid = "0.5"
md_version = "0.1"
md_name = "Sleep"
md_description = "Sleep demo"
md_license = "BSD-2"

class Plugin(QueryHandler):
    def id(self):
        return __name__

    def name(self):
        return md_name

    def description(self):
        return md_description

    def defaultTrigger(self):
        return 'sle '

    def handleQuery(self, query):
        time.sleep(1) # <---- 1 second for a better demonstration
        s = query.string.strip().lower()
        query.add(Item(
            id=__name__,
            text=s,
            actions=[Action(
                id='copy',
                text='Copy to clipboard',
                callable=lambda: setClipboardText(s)
            )]
        ))
```

1. Type "sle test"
2. Only "sle t" will be visible in the textbox, in logs we have something like:
```
23:33:05 [warn:albert] Busy wait on query: #26
23:33:06 [warn:albert] Busy wait on query: #27
23:33:07 [warn:albert] Busy wait on query: #28
```
3. After some delay "sle t" is replaced by "sle test"
4. After 1 second I see an item in the menu as expected

# asyncio demo

```python
# ~/.local/share/albert/python/plugins/asyncio-demo/__init__.py
from albert import *

import asyncio
import threading

md_iid = "0.5"
md_version = "0.1"
md_name = "Asyncio"
md_description = "Asyncio demo"
md_license = "BSD-2"

class Plugin(QueryHandler):
    is_working = True  # For a background task, that makes the EventLoop alive
    query = None  # The current query
    thead = None  # We need this to move work to a background
    loop = None  # To submit tasks and cancel 
    curr_task = None  # To cancel when a new symbol is entered and we haven't yet made the job

    def id(self):
        return __name__

    def name(self):
        return md_name

    def description(self):
        return md_description

    def defaultTrigger(self):
        return 'aio '

    def initialize(self):
        self.loop = asyncio.get_event_loop()
        self.thread = threading.Thread(target=self.runLoop)
        self.thread.start()

    def finalize(self):
        self.is_working = False
        if self.curr_task:
            self.curr_task.cancel()
        self.thread.join()

    def handleQuery(self, query):
        if self.curr_task:
            self.curr_task.cancel()
        self.query = query
        self.curr_task = self.loop.create_task(self.debounced())

    # Internal

    async def debounced(self):
        await asyncio.sleep(1)  # <---- waiting 1 second too
        if not self.query.isValid:
            return

        s = self.query.string.strip().lower()
        self.query.add(Item(
            id=__name__,
            text=s,
            actions=[Action(
                id='copy',
                text='Copy to clipboard',
                callable=lambda: setClipboardText(s)
            )]
        ))

    def runLoop(self):
        self.loop.run_until_complete(self.watchLoop())

    async def watchLoop(self):
        while self.is_working:
            await asyncio.sleep(1)
```

1. Type "aio test"
2. The typing is not blocked and run smoothly. Also, I didn't see warnings in the log.
3. After 1 second I see an item in the menu as expected

**Comment by ManuelSchneid3r** [comment]
Ah now I got it. Well several problems: 
2. I have nothing to cite but I feels like mixing real parallelization and serialization of coroutines is nothing we(one) should do.
3. The problem is artificially because blocking the thread makes no sense. Plugin devs could easily timeout every few ms and release the thread asap if requested. Also * (later). But I see your point plugin devs _can_ break UX and it would be nice if the could not. Unfortunately the approach above does not help here. The async component had to go to the python (c++) plugin somehow and the handleQuery 
4. Ownership. The query is borrowed. C++ is not joking about memory and ownership. If the query is finished the frontend is free to delete it. How would the Query object know when it is okay not to busy wait in the destructor? Asyncio is threaded as well and imho it gets quite complicated meanwhile. The main/gui thread, multiple threads _per_ query, multiple parallel queries, multiple plugins with indexing threads,... now asyncio. I am not sure if I like this complexity. 

Let's keep it simple

**Comment by ManuelSchneid3r** [comment]
Oops still editing

**Comment by ManuelSchneid3r** [comment]
Done

**Comment by vsuharnikov** [comment]
I use `asyncio` not because I want to use it, but because a solution with `thread.sleep` has mentioned UX issues :) 
BTW I've found how to get rid of `asyncio` after the reading your last message, see below (A).

> Plugin devs could easily timeout every few ms and release the thread asap if requested

This code:
```python
for number in range(50):
    sleep(0.01)
    if not query.isValid:
        return

job()
```
works without issues if `job()` is fast (except warnings in the log). 

But if `job()` runs for example a slow external process:
1. We can't insert checks `query.isValid` inside the external process to cancel it (but there is a workaround (A)).
2. It leads to input freezes.

> The async component had to go to the python (c++) plugin somehow

You mean to make the process of input independent on `handleQuery`? This could help.

# (A) interruptable threads demo

```python
from albert import *
import time
import threading

md_iid = "0.5"
md_version = "0.1"
md_name = "Thread"
md_description = "Thread demo"
md_license = "BSD-2"


class Plugin(QueryHandler):
    job_thread = None
    last_query = None

    def id(self):
        return __name__

    def name(self):
        return md_name

    def description(self):
        return md_description

    def defaultTrigger(self):
        return 'th '

    def finalize(self):
        if self.job_thread:
            self.job_thread.stop()

    def handleQuery(self, query):
        # cancel if there is a new query
        for number in range(50):
            time.sleep(0.01)
            if not query.isValid:
                return

        # input freezes without a thread, try to hold on "y" to see freezes:
        # self.job()

        self.last_query = query
        if self.job_thread:
            self.job_thread.stop()
        self.job_thread = InterruptableThread(target=self.job)
        self.job_thread.start()

    # Do not copy last_query to a local variable, otherwise you will get a segfault!
    def job(self):
        s = self.last_query.string.strip().lower()
        if s == '':
            return

        # Complex calculations
        time.sleep(1)  # <---- 1 second for a better demonstration

        item = Item(
            id=__name__,
            text=s,
            actions=[Action(
                id='copy',
                text='Copy to clipboard',
                callable=lambda: setClipboardText(s)
            )]
        )

        if self.last_query.isValid and not threading.current_thread().stopped():
            self.last_query.add(item)


# https://www.geeksforgeeks.org/python-different-ways-to-kill-a-thread/
class InterruptableThread(threading.Thread):
    def __init__(self, *args, **kwargs):
        super(InterruptableThread, self).__init__(*args, **kwargs)
        self._stop = threading.Event()

    def stop(self):
        self._stop.set()

    def stopped(self):
        return self._stop.isSet()
```

**Comment by ManuelSchneid3r** [comment]
> You mean to make the process of input independent on `handleQuery`? This could help.

I mean that the loop would be instanciated by the python c++ plugin and the handlequery is declare async. Still theres the problem of ownership of the query object.

How is `self._stop` in your example intended to be used. If I understand correctly you would have to check `self._stop` too. Thats equivalent to `query.isValid`.

**Comment by ManuelSchneid3r** [comment]
just released a fix which should render this one obsolete. busy wait on queries in the frontend is now postponed until the user closes the app. the ui should be fluid again.
